src/lib/comms/sol-netctl-impl-connman.c

```diff
@@ -1212,3 +1222,274 @@ sol_netctl_service_state_to_str(enum sol_netctl_service_state state)
+
+    return 0;
+error:
+    sd_bus_reply_method_return(m, NULL);
+    return -EINVAL;
+}
+
+static int
+_agent_input_properties(sd_bus_message *m, struct ctx *context)
+{
+    enum sol_netctl_agent_prompt_type type;
+    const struct sol_netctl_agent *agent = context->agent;
+    char *str;
+    struct sol_netctl_agent_input *input;
+    int r;
+
+    r = sd_bus_message_enter_container(m, SD_BUS_TYPE_ARRAY, "{sv}");
+    SOL_INT_CHECK(r, < 0, r);
+
+    _destory_agent_vector(&_ctx.agent_vector);
+    sol_ptr_vector_init(&_ctx.agent_vector);
+
+    do {
+        static const struct sol_str_table table[] = {
+            SOL_STR_TABLE_ITEM("Name",
+                SOL_NETCTL_AGENT_NAME),
+            SOL_STR_TABLE_ITEM("Identity",
+                SOL_NETCTL_AGENT_IDENTITY),
+            SOL_STR_TABLE_ITEM("Passphrase",
+                SOL_NETCTL_AGENT_PASSPHRASE),
+            SOL_STR_TABLE_ITEM("WPS",
+                SOL_NETCTL_AGENT_WPS),
+            { }
+        };
+
+        r = sd_bus_message_enter_container(m, SD_BUS_TYPE_DICT_ENTRY, "sv");
+        SOL_INT_CHECK_GOTO(r, < 1, end);
+
+        str = NULL;
+        r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &str);
+        SOL_INT_CHECK(r, < 0, r);
+
+        if (str)
+            type = sol_str_table_lookup_fallback(table,
+                sol_str_slice_from_str(str),
+                SOL_NETCTL_AGENT_UNKNOWN);
+        else
+            type = SOL_NETCTL_AGENT_UNKNOWN;
+
+        if (type != SOL_NETCTL_AGENT_UNKNOWN) {
+            int is_exist = false;
+            int i = 0;
+
+            SOL_PTR_VECTOR_FOREACH_IDX (&_ctx.agent_vector, input, i) {
+                if (input->type == type) {
+                    is_exist = true;
+                    break;
+                }
+            }
+
+            if (is_exist == false) {
+                input = calloc(1, sizeof(struct sol_netctl_agent_input));
+                SOL_NULL_CHECK(input, -ENOMEM);
+                input->type = type;
+                r = sol_ptr_vector_append(&_ctx.agent_vector, input);
+                SOL_INT_CHECK(r, < 0, r);
+            }
+        }
+
+        r = sd_bus_message_skip(m, "v");
+        SOL_INT_CHECK(r, < 0, r);
+
+        r = sd_bus_message_exit_container(m);
+        SOL_INT_CHECK(r, < 0, r);
+    } while (1);
+
+end:
+    if (r == 0) {
+        r = sd_bus_message_exit_container(m);
+        SOL_INT_CHECK(r, < 0, r);
+
+        if (sol_ptr_vector_get_len(&_ctx.agent_vector) > 0) {
+            agent->request_input((void *)context->agent_data, context->auth_service,
+                (const struct sol_ptr_vector *)&_ctx.agent_vector);
+        }
+    }
+
+    return r;
+}
+
+static int
+agent_report_input(sd_bus_message *m, void *data, sd_bus_error *ret_error)
+{
+    struct sol_netctl_service *service;
+    const char *path;
+    struct ctx *context = (struct ctx *)data;
+    const struct sol_netctl_agent *agent = context->agent;
+    int r;
+
+    SOL_NULL_CHECK_GOTO(agent, error);
```

Comment:
I don't think this check is necessary. These callback will only be valid if agent is NOT null anyway. You already did this check in sol_netctl_register_agent